apps/prairielearn/elements/pl-file-upload/pl-file-upload.py

```diff
@@ -87,36 +220,54 @@ def render(element_html: str, data: pl.QuestionData) -> str:
 
 def parse(element_html: str, data: pl.QuestionData) -> None:
     element = lxml.html.fragment_fromstring(element_html)
-    raw_file_names = pl.get_string_attrib(element, "file-names", "")
-    required_file_names = get_file_names_as_array(raw_file_names)
-    answer_name = get_answer_name(raw_file_names)
+    allow_blank = pl.get_boolean_attrib(element, "allow-blank", ALLOW_BLANK_DEFAULT)
+    raw_required_file_names = pl.get_string_attrib(element, "file-names", "")
+    required_file_names = get_file_names_as_array(raw_required_file_names)
+    raw_optional_file_names = pl.get_string_attrib(element, "optional-file-names", "")
+    optional_file_names = get_file_names_as_array(raw_optional_file_names)
+    answer_name = get_answer_name(raw_required_file_names, raw_optional_file_names)
 
-    # Get submitted answer or return parse_error if it does not exist
-    files = data["submitted_answers"].get(answer_name, None)
-    if not files:
-        add_format_error(data, "No submitted answer for file upload.")
+    # Get submitted answer or return format error if it does not exist
+    files = data["submitted_answers"].get(answer_name)
+    if not files and not allow_blank:
+        add_format_error(answer_name, data, "No submitted answer for file upload.")
         return
 
     # We will store the files in the submitted_answer["_files"] key,
-    # so delete the original submitted answer format to avoid
-    # duplication
+    # so delete the original submitted answer format to avoid duplication
     del data["submitted_answers"][answer_name]
 
     try:
-        parsed_files = json.loads(files)
+        parsed_files = json.loads(files)  # pyright: ignore [reportArgumentType]
     except Exception:
-        add_format_error(data, "Could not parse submitted files.")
+        add_format_error(answer_name, data, "Could not parse submitted files.")
         parsed_files = []
 
-    # Filter out any files that were not listed in file_names
-    parsed_files = [x for x in parsed_files if x.get("name", "") in required_file_names]
+    # Split optional names into two separate lists: wildcard patterns and plain names
+    optional_files_regex, optional_files_plain = extract_patterns(optional_file_names)
+
+    wildcard_files = find_matching_files(
+        optional_files_regex, [x.get("name", "") for x in parsed_files]
+    )
+
+    parsed_files = [
+        x
+        for x in parsed_files
+        if x.get("name", "") in required_file_names
```

Comment:
The problem with this and the suggestion above is that in the later steps the code needs a list of files, not just their names. So unless I'm missing something here, I will still need to do another pass to filter the files based on the accepted names in the end. Here's how it would look like, not sure if that's actually an improvement compared to how it is currently...

```  
    parsed_file_names = [x.get("name", "") for x in parsed_files]

    wildcard_files = find_matching_files(
        optional_files_regex, parsed_file_names
    )
    
    accepted_file_names = parsed_file_names & (required_file_names | optional_files_plain | wildcard_files)

    parsed_files = [x for x in parsed_files if x.get("name", "") in accepted_file_names]```

Reply:
Ahh that makes sense, I thought there might have been some context I was missing. Feel free to ignore and mark off this item.